File: HelloTuiCpp/Classes/scene/Bag/Bagui.cpp
#include "Bagui.h"
#include "tui/tagMap/Tag_bag.h"
#include "tui/TuiManager.h"

NS_BAG_BEGIN

void Bagui::onLoadScene()
{
	setCachable(true);

	SpriteFrameCache::getInstance()->addSpriteFramesWithFile("bag/bagui.plist");
	TuiManager::getInstance()->parseScene(this, "panel_bag", PATH_BAG);

	CGridView* pGridView = (CGridView*)this->getControl(PANEL_BAG, GV_BAG);
	pGridView->setDataSourceAdapter(this, ccw_datasource_adapter_selector(Bagui::event_adapt_gvBag));
	pGridView->reloadData();

	CTableView* pTblView = (CTableView*)this->getControl(PANEL_BAG, TBL_BAG);
	pTblView->setDataSourceAdapter(this, ccw_datasource_adapter_selector(Bagui::event_adapt_tblBag));
	pTblView->reloadData();

	CPageView* pPageView = (CPageView*)this->getControl(PANEL_BAG, PV_BAG);
	pPageView->setDataSourceAdapter(this, ccw_datasource_adapter_selector(Bagui::event_adapt_pageBag));
	pPageView->reloadData();

	CButton* pBtnView = (CButton*)this->getControl(PANEL_BAG, BTN_BACK);
	pBtnView->setOnClickListener(this, ccw_click_selector(Bagui::event_btn_back));
}

Ref* Bagui::event_adapt_gvBag(Ref* pConvertCell, unsigned int uIdx)
{
	CGridViewCell* pCell = (CGridViewCell*)pConvertCell;
	CButton *btn = nullptr;
	if (!pCell)
	{
		pCell = new CGridViewCell();
		pCell->autorelease();

		TuiManager::getInstance()->parseCell(pCell, "cell_grid", PATH_BAG);

		btn = (CButton*)pCell->getChildByTag(BTN_GRID);
		btn->setOnClickListener(this, ccw_click_selector(Bagui::event_btn_gv));
	}else{
		btn = (CButton*)pCell->getChildByTag(BTN_GRID);
	}
	btn->setUserTag(uIdx);
	return pCell;
}

Ref* Bagui::event_adapt_tblBag(Ref* pConvertCell, unsigned int uIdx)
{
	CTableViewCell* pCell = (CTableViewCell*)pConvertCell;
	if (!pCell)
	{
		pCell = new CTableViewCell();
		pCell->autorelease();

		TuiManager::getInstance()->parseCell(pCell, "cell_grid", PATH_BAG);
	}else{

	}
	return pCell;
}

Ref* Bagui::event_adapt_pageBag(Ref* pConvertCell, unsigned int uIdx)
{
	CPageViewCell *pCell = (CPageViewCell*)pConvertCell;
	if (!pCell)
	{
		pCell = new CPageViewCell();
		pCell->autorelease();

		TuiManager::getInstance()->parseCell(pCell, "cell_page", PATH_BAG);
	}else{

	}
	return pCell;
}

void Bagui::event_btn_back(Ref* pSender)
{
	/*
	CSceneManager::getInstance()->replaceScene(
		CCSceneExTransitionFade::create(0.5f, LoadScene("Main::Mainui")));
		*/
	CSceneManager::getInstance()->popUIScene(this);
}

void Bagui::event_btn_gv(Ref* pSender)
{
	CButton *btn = (CButton*)pSender;
	int id = btn->getUserTag();

	CCLOG("id = %d", id);
}

/************************************************************************/
//	GET/SET/IS
/************************************************************************/
Node* Bagui::getPanel( int tagPanel )
{
    Node *pPanel = nullptr;
    switch (tagPanel)
    {
        case PANEL_BAG:
            pPanel = this->getChildByTag(PANEL_BAG);
            break;
    }
    return pPanel;
}

NS_BAG_END

File: HelloTuiLua/src/bag/bagui.lua
require "extern"
require "tagMap/Tag_bagui"

Bagui = class("Bagui",function()
	return TuiBase:create()
end)

Bagui.__index = Bagui
local __instance = nil

function Bagui:create()
	local ret = Bagui.new()
	__instance = ret
	ret:setOnLoadSceneScriptHandler(function() ret:onLoadScene() end)
	return ret
end

function Bagui:getControl(tagPanel,tagControl)
	local ret = nil
	ret = self:getPanel(tagPanel):getChildByTag(tagControl)
	return ret
end

function Bagui:getPanel(tagPanel)
	local ret = nil
	if tagPanel == Tag_bagui.PANEL_BAG then
		ret = self:getChildByTag(tagPanel)
	end
	return ret
end

---------------logic----------------------------

local function event_btn_back(p_sender)
	
	CSceneManager:getInstance():replaceScene(
		CCSceneExTransitionFade:create(0.5,LoadScene("Welcomeui")))
end

local function event_btn_recombine(p_sender)

	CSceneManager:getInstance():replaceScene(
		CCSceneExTransitionFade:create(0.5,LoadScene("Recombineui")))
end

local function event_btn_click(p_sender)
	print("tag : "..tolua.cast(p_sender,"ccw.CWidget"):getUserTag())
end

local function event_adapt_gvbag(p_convertview, idx)
	local pCell = p_convertview
	local btn = nil
	if pCell == nil then
		pCell = CGridViewCell:new()

		TuiManager:getInstance():parseCell(pCell, "cell_grid", PATH_BAGUI)
		btn = pCell:getChildByTag(Tag_bagui.BTN_GRID)
		btn:setOnClickScriptHandler(event_btn_click)
	else
		btn = pCell:getChildByTag(Tag_bagui.BTN_GRID)
	end
	tolua.cast(btn,"ccw.CWidget"):setUserTag(idx)
	return pCell
end

local function event_adapt_tblbag(p_convertview, idx)
	local pCell = p_convertview
	if pCell == nil then
		pCell = CTableViewCell:new()

		TuiManager:getInstance():parseCell(pCell, "cell_grid", PATH_BAGUI)
	else

	end
	return pCell
end
local function event_adapt_pvbag(p_convertview, idx)
	local pCell = p_convertview
	if pCell == nil then
		pCell = CPageViewCell:new()

		TuiManager:getInstance():parseCell(pCell, "cell_page", PATH_BAGUI)
	else

		 
	end
	pCell:getChildByTag(Tag_bagui.BTN_GRID4):setOnClickScriptHandler(function(sender)

	 		CSceneManager:getInstance():runUIScene(LoadScene("MsgBox"))
	 end)
	return pCell
end

function Bagui:onLoadScene()
	-- cc.SpriteFrameCache:getInstance():addSpriteFrames("bag/bagui.plist")
	TuiManager:getInstance():parseScene(self,"panel_bag",PATH_BAGUI)

	local btnBack = self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.BTN_BACK)
	btnBack:setOnClickScriptHandler(event_btn_back)
	
	local btnRecombine = self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.BTN_RECOMBINE)
	btnRecombine:setOnClickScriptHandler(event_btn_recombine)
	
	local gvBag = self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.GV_BAG)
	gvBag:setDataSourceAdapterScriptHandler(event_adapt_gvbag)
	gvBag:reloadData()
	
	local tblBag = self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.TBL_BAG)
	tblBag:setDataSourceAdapterScriptHandler(event_adapt_tblbag)
	tblBag:reloadData()
	
	local pvBag = self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.PV_BAG)
	pvBag:setDataSourceAdapterScriptHandler(event_adapt_pvbag)
	pvBag:reloadData()
	
end